glide-core/redis-rs/redis/src/cluster_async/connections_container.rs

```diff
@@ -134,11 +140,163 @@ impl<Connection> std::fmt::Display for ConnectionsMap<Connection> {
     }
 }
 
+#[derive(Clone, Debug)]
+pub(crate) struct RefreshTaskNotifier {
+    notify: Arc<Notify>,
+}
+
+impl RefreshTaskNotifier {
+    fn new() -> Self {
+        RefreshTaskNotifier {
+            notify: Arc::new(Notify::new()),
+        }
+    }
+
+    pub fn get_notifier(&self) -> Arc<Notify> {
+        self.notify.clone()
+    }
+
+    pub fn notify(&self) {
+        self.notify.notify_waiters();
+    }
+}
+
+// Enum representing the task status during a connection refresh.
+//
+// - **Reconnecting**:
+//   Indicates that a refresh task is in progress. This status includes a dedicated
+//   notifier (`RefreshTaskNotifier`) so that other tasks can wait for the connection
+//   to be refreshed before proceeding.
+//
+// - **ReconnectingTooLong**:
+//   Represents a situation where a refresh task has taken too long to complete.
+//   The status transitions from `Reconnecting` to `ReconnectingTooLong` under specific
+//   conditions (e.g., after one attempt of reconnecting inside the task or after a timeout).
+//
+// When transitioning from `Reconnecting` to `ReconnectingTooLong`, the associated
+// notifier is triggered to unblock all awaiting tasks.
+#[derive(Clone, Debug)]
+pub(crate) enum RefreshTaskStatus {
+    // The task is actively reconnecting. Includes a notifier for tasks to wait on.
+    Reconnecting(RefreshTaskNotifier),
+    // The task has exceeded the allowed reconnection time.
+    #[allow(dead_code)]
+    ReconnectingTooLong,
+}
+
+impl Drop for RefreshTaskStatus {
+    fn drop(&mut self) {
+        if let RefreshTaskStatus::Reconnecting(notifier) = self {
+            debug!("RefreshTaskStatus: Dropped while in Reconnecting status. Notifying tasks.");
+            notifier.notify();
+        }
+    }
+}
+
+impl RefreshTaskStatus {
+    /// Creates a new `RefreshTaskStatus` in the `Reconnecting` status with a fresh `RefreshTaskNotifier`.
+    pub fn new() -> Self {
+        debug!("RefreshTaskStatus: Initialized in Reconnecting status with a new notifier.");
+        RefreshTaskStatus::Reconnecting(RefreshTaskNotifier::new())
+    }
+
+    // Transitions the current status from `Reconnecting` to `ReconnectingTooLong` in place.
+    //
+    // If the current status is `Reconnecting`, this method notifies all waiting tasks
+    // using the embedded `RefreshTaskNotifier` and updates the status to `ReconnectingTooLong`.
+    //
+    // If the status is already `ReconnectingTooLong`, this method does nothing.
+    #[allow(dead_code)]
+    pub fn flip_status_to_too_long(&mut self) {
+        if let RefreshTaskStatus::Reconnecting(notifier) = self {
+            debug!(
+                "RefreshTaskStatus: Notifying tasks before transitioning to ReconnectingTooLong."
+            );
+            notifier.notify();
+            *self = RefreshTaskStatus::ReconnectingTooLong;
+        } else {
+            debug!("RefreshTaskStatus: Already in ReconnectingTooLong status.");
```

Comment:
return an error that would indicate that something unexpected happened (you can also postpone these functions to the next PR)